Fix testsuite when configured with -safe-string

dra27 opened the pull request:
Quite a large number of tests were failing when OCaml is configured with `-safe-string`. This commit fixes all of those tests - tested with/without -flambda and with/without -safe-string on Ubuntu 14.04 and Windows 7 Pro/mingw

dra27 committed:
Fix testsuite when configured with -safe-string

dra27 commented on testsuite/tests/basic/patmatch.ml:
```diff
@@ -149,8 +149,8 @@ let () =
-  | true -> "a"
-  | false -> "a"
+  | true -> Bytes.of_string "a"
+  | false -> Bytes.of_string "a"
```
@maranget, @chambart - does adding `Bytes.of_string` here completely defeat the purpose of this test?

damiendoligez replied:
I think it does, but the test is rather pointless in `-safe-string` mode anyway.

dra27 replied:
In which case, shall I split that test out to a separate file and do a similar shim to https://github.com/ocaml/ocaml/commit/fa244ddaea2aa258fc8f4e38ac0768affd2f31c5 to put a dummy ML file for `-safe-string` mode?

damiendoligez replied:
That would be good, yes.

dra27 commented on testsuite/tests/typing-safe-linking/Makefile:
```diff
@@ -3,6 +3,13 @@
 
+default:
+	@if $(SAFE_STRING) ; then \
+	  echo " ... testing 'b_bad.ml' => skipped" ; \
```
@garrigue - I think it's correct that since `-unsafe-string` is unavailable when OCaml is configured with `-safe-string` that this test should be completely skipped, as it's not possible to build `a.cmo` with mutable strings?

damiendoligez replied:
You could remove the definition of `A.y` and the last line of `b_bad.ml` because the program is never executed anyway, then you could compile in `-safe-string` mode. What I don't understand is that I get a warning 8 with or without the `-safe-string` option. @garrigue ?

garrigue replied:
This is precisely the point: `-safe-string` should not discard the warning, as we could create  a `Y` of type `string t` in another module.

dra27 replied:
Does that mean that if OCaml is configured with `-safe-string`, therefore, that it's still worth running this test, just with the bottom line of both `a.ml` and `b_bad.ml` chopped?

damiendoligez replied:
I'd comment them out and add a note.

damiendoligez commented:
@dra27 Do you have the time to do the suggested changes?

dra27 commented:
Yes - I shall try to do this tomorrow

dra27 commented:
(that's just a rebase, I'll put the changes in squashable commits as I do them...)

dra27 committed:
Proposed fixup for tests/typing-safe-linking

Disables the A.y test when OCaml configured with -safe-string

dra27 commented:
First fix (do not merge yet - I haven't run a full platform test)

dra27 committed:
Proposed fixup for tests/basic/patmatch.ml

Moves the PR6322 test to a separate file and disables it when OCaml
configured with -safe-string

dra27 commented:
Right - two commits fixing up those two issues. I have done some very basic testing. I'm going to rebase #807, especially as there have been some more tests added. When I run the tests, I'll include this branch locally to test it fully.

gasche commented:
Tell me once you have a specific branch (or git commit) that is ready for testing, I can run INRIA's Continuous Integration testing on it.

dra27 commented:
@gasche - thanks! Which Windows configurations are tested on it?

shindere commented:
There are six machines: msvc, mingw and cygwin in both 32 and 64 bits
flavours.

dra27 commented:
And do they just test vanilla, or flambda/safe-string combinations?
/cc @avsm just to be in mind (if it's not already) for Azure stuff later this year...

dra27 commented:
@gasche - https://github.com/dra27/ocaml/tree/testsuite-testing ... I'm just about to start running tests on it here

shindere commented:
David Allsopp (2016/10/19 06:14 -0700):

> And do they just test vanilla, or flambda/safe-string combinations?

From what I could see for the cygwin-32 machine on the trunk branch,
just vanilla.

At least on this machine, tests are run by calling the tools/ci-build
script from the OCaml git repository with no argument except
architecture and directory.

damiendoligez commented:
> And do they just test vanilla, or flambda/safe-string combinations?

They don't do flambda nor safe-string at this time.

I can easily reconfigure some of the CI jobs to test one or the other (or both). What I cannot do is allocate enough machines to test all combinations of toolchain/flambda/safe-string.

Should I just set these options on a random subset?

damiendoligez commented:
> At least on this machine, tests are run by calling the tools/ci-build
> script from the OCaml git repository with no argument except
> architecture and directory.

Minor caveat: they all use the ci-build script from trunk, even when testing some other branch.

dra27 commented:
@damiendoligez - probably best not to change it for now? There'll be more on this anon, I'm just making mental notes to check this and not unnecessarily duplicate efforts! Windows CI is high (not just for the compiler) is high on my OCL list...

shindere commented:
Damien Doligez (2016/10/19 07:09 -0700):

> > And do they just test vanilla, or flambda/safe-string combinations?
> 
> They don't do flambda nor safe-string at this time.
> 
> I can easily reconfigure some of the CI jobs to test one or the other (or both). What I cannot do is allocate enough machines to test all combinations of toolchain/flambda/safe-string.
> 
> Should I just set these options on a random subset?

Would it make sense to do some rotation on the optons for one given
platform? So one build would be vanila, the next one flambda, then
safe-strings, then both, then again vanilla...

Or perhaps all the non-vanilla configurations could be tested only once
a day or so?

mshinwell commented:
Testing flambda + safe-string is probably more important than testing flambda without safe-string at this stage, I suspect?  (Just to eliminate one combination...)

dra27 commented:
Something running less frequently than full CI but reporting on breakages on trunk (or other important branches) was how I'm aiming to monitor the support for very old Windows versions and very old Microsoft C Compilers. Seems like a good idea for some of the stranger build configurations too.

dra27 commented:
Right, this and #807 are now I hope finalised and ready for review (let me know if the two fixup commits here are OK, and I'll squash them).
My laptop is presently screaming as it builds 16 Windows builds in a VM...

https://github.com/dra27/ocaml/tree/testsuite-testing is up-to-date if anyone wants to test both PRs simultaneously, but I'm not anticipating any errors this time round (famous last words...)

shindere commented:
Mark Shinwell (2016/10/20 23:03 -0700):

> Testing flambda + safe-string is probably more important than testing
> flambda without safe-string at this stage, I suspect?  (Just to
> eliminate one combination...)

Thanks. We'll see whether that can be added.

damiendoligez commented:
The fixups are OK. No need to squash, we can do it from GitHub's GUI.
